### PR TITLE
External libraries

### PR DESCRIPTION
**Pull request by etokareva**

### CODE REVIEW

File: src/plugins/export/export.js [hunk 1]
```diff
@@ -24,44 +25,79 @@ class Export extends Plugin {
...
+				if (!pdfLib || !autotableLib) {
+					throw new AppError('jsPDF', 'To use export plugin for pdf format please add https://github.com/MrRio/jsPDF and https://github.com/simonbengtsson/jsPDF-AutoTable libraries to your project');
+				}
+				const pdf = new Pdf(pdfLib);
 				pdf.write(this.rows, this.columns, this.id);
 			}
 		});
 	}
 
+	get imports() {
```
Comment (klumba12) [inline src/plugins/export/export.js]: better approach:
importLib(name){
if (not found) throw error

return lib;
}

File: src/plugins/export/pdf.js [hunk 2]
```diff
@@ -1,12 +1,11 @@
-import {AppError} from '@grid/core/infrastructure';
 import {flatView} from '@grid/core/export/export.service';
 
 export class Pdf {
+	constructor(lib) {
```
Comment (klumba12) [inline src/plugins/export/pdf.js]: rename lib to jsPDF

File: src/plugins/export/xlsx.js [hunk 3]
```diff
@@ -15,21 +14,12 @@ function toArrayBuffer(excel) {
...
-	for (let i = range.s.r; i <= range.e.r; ++i) {
-		const address = XLSX.utils.encode_col(i) + '1';
-		if (!worksheet[address]) continue;
-		worksheet[address].v = headers[i];
-	}
-	return worksheet;
-}
+
 export class Xlsx {
+	constructor(lib) {
```
Comment (klumba12) [inline src/plugins/export/xlsx.js]: rename lib to xlxs

File: src/plugins/export/xlsx.js [hunk 4]
```diff
@@ -15,21 +14,12 @@ function toArrayBuffer(excel) {
...
+		this.XLSX = lib;
+	}
```
Comment (klumba12) [inline src/plugins/export/xlsx.js]: consider to use this.xlsx

File: src/plugins/export/download.js [hunk 5]
```diff
@@ -1,9 +1,4 @@
-import {AppError} from '@grid/core/infrastructure';
-
-export function download(name, data, mimeType, extension) {
-	if (!window.fileSaver) {
-		throw new AppError('fileSaver', 'To use export plugin for pdf format please add https://github.com/eligrey/FileSaver.js library to your project');
-	}
+export function download(fileSaver, name, data, mimeType, extension) {
```
Comment (klumba12) [inline src/plugins/export/download.js]: add factory here
downloadFactory(fileSaver){
return (name, data, ...)

File: demo/pages/export/index.js [hunk 6]
```diff
@@ -1,7 +1,24 @@
...
+import jsPDF from 'jspdf';
+import autotable from 'jspdf-autotable';
+
+Controller.$inject = ['$http', 'qgrid'];
+export default function Controller($http, qgrid) {
 	this.rows = [];
 
+	this.model = qgrid.model();
+	this.model.plugin({
+		imports: {
```
Comment (vkorolev) [inline demo/pages/export/index.js:12]: I don't think it's a good idea to require external dependencies like this. Why not use commonjs?

File: src/plugins/export/export.js [hunk 7]
```diff
@@ -24,44 +25,75 @@ class Export extends Plugin {
...
 				pdf.write(this.rows, this.columns, this.id);
 			}
 		});
 	}
 
+	importLib(name) {
+		const lib = this.model.plugin().imports[name];
+		if (!lib) {
+			switch (name) {
+				case 'xlsx': {
```
Comment (klumba12) [inline src/plugins/export/export.js]: AppError('export', '...') for every error

File: src/plugins/export/export.js [hunk 8]
```diff
@@ -24,44 +25,75 @@ class Export extends Plugin {
...
+					throw new AppError('xlsx', 'To use export plugin for xlsx format please add http://github.com/SheetJS/js-xlsx library to your project');
+				}
+				case 'fileSaver': {
+					throw new AppError('fileSaver', 'To use export plugin for file saving please add https://github.com/eligrey/FileSaver.js library to your project');
+				}
+				case 'pdf': {
+					throw new AppError('jsPDF', 'To use export plugin for pdf format please add https://github.com/MrRio/jsPDF and https://github.com/simonbengtsson/jsPDF-AutoTable libraries to your project');
+				}
+				default: {
+					throw new AppError('import library', 'No such library in imports');
```
Comment (klumba12) [inline src/plugins/export/export.js]: `Can't find ${name} library in imports'

File: src/plugins/import/read.js [hunk 9]
```diff
@@ -107,6 +108,26 @@ function readFile(e, file, model, options = {}) {
 		}
 	}
 }
+function importLib(model, name) {
```
Comment (klumba12) [inline src/plugins/import/read.js]: rename importLib to resolveLib... import looks confusing a little bit :)

File: src/core/plugin/plugin.service.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
+import {AppError} from '@grid/core/infrastructure';
+
+export function resolve(model, name) {
```
Comment (klumba12) [inline src/core/plugin/plugin.service.js]: please Make it as a class
class PluginService{
constructor(model);
resolve(name);
}

**Review by klumba12 — APPROVED** [review]

**Pull request merged**